test/unit-tests.c

```diff
@@ -356,6 +358,98 @@ test_attach_ctx(void **state __attribute__((unused)))
     assert_int_equal(-1, vfu_attach_ctx(&vfu_ctx));
 }
 
+/*
+ * FIXME expand and validate
+ */
+static void
+test_vfu_ctx_create(void **state __attribute__((unused)))
+{
+    vfu_ctx_t *vfu_ctx = NULL;
+    vfu_pci_hdr_id_t id = { 0 };
+    vfu_pci_hdr_ss_t ss = { 0 };
+    vfu_pci_hdr_cc_t cc = { 0 };
+    vfu_cap_t pm = {.pm = {.hdr.id = PCI_CAP_ID_PM}};
+    vfu_cap_t *caps[] = { &pm };
+
+    vfu_ctx = vfu_create_ctx(VFU_TRANS_SOCK, "", LIBVFIO_USER_FLAG_ATTACH_NB,
+                             NULL, VFU_DEV_TYPE_PCI);
+    assert_non_null(vfu_ctx);
+    assert_int_equal(0,
+                     vfu_pci_setup_config_hdr(vfu_ctx, id, ss, cc,
+                                              VFU_PCI_TYPE_CONVENTIONAL, 0));
+    assert_int_equal(0, vfu_pci_setup_caps(vfu_ctx, caps, 1));
+    assert_int_equal(0, vfu_realize_ctx(vfu_ctx));
+}
+
+static void
+test_pci_caps(void **state __attribute__((unused)))
+{
+    vfu_pci_config_space_t config_space;
+    vfu_ctx_t vfu_ctx = { .pci.config_space = &config_space };
+    vfu_cap_t pm = {.pm = {.hdr.id = PCI_CAP_ID_PM, .pmcs.raw = 0xabcd }};
+    vfu_cap_t *vsc[2] = {
+        alloca(sizeof(struct vsc) + 5),
+        alloca(sizeof(struct vsc) + 13)
+    };
+    vfu_cap_t *vfu_caps[] = { &pm, vsc[0], vsc[1] };
+    struct caps *caps;
+    int err;
+    struct pmcap pmcap = { .pmcs.raw = 0xef01 };
+    off_t off;
+
+    vsc[0]->vsc.hdr.id = PCI_CAP_ID_VNDR;
+    vsc[0]->vsc.size = 8;
+    memcpy(vsc[0]->vsc.data, "abcde", 5);
+
+    vsc[1]->vsc.hdr.id = PCI_CAP_ID_VNDR;
+    vsc[1]->vsc.size = 16;
+    memcpy(vsc[1]->vsc.data, "Hello world.", 12);
+
+    caps = caps_create(&vfu_ctx, vfu_caps, 3, &err);
+    assert_non_null(caps);
+
+    /* check that capability list is placed correctly */
+    assert_int_equal(PCI_CAP_ID_PM,
+                     config_space.raw[PCI_STD_HEADER_SIZEOF + PCI_CAP_LIST_ID]);
+    assert_int_equal(PCI_STD_HEADER_SIZEOF + PCI_PM_SIZEOF,
+                     config_space.raw[PCI_STD_HEADER_SIZEOF + PCI_CAP_LIST_NEXT]);
+    assert_int_equal(PCI_CAP_ID_VNDR,
+                     config_space.raw[PCI_STD_HEADER_SIZEOF + PCI_PM_SIZEOF + PCI_CAP_LIST_ID]);
+    assert_int_equal(PCI_STD_HEADER_SIZEOF + PCI_PM_SIZEOF + vsc[0]->vsc.size,
+                     config_space.raw[PCI_STD_HEADER_SIZEOF + PCI_PM_SIZEOF + PCI_CAP_LIST_NEXT]);
+    assert_int_equal(8,
+                     config_space.raw[PCI_STD_HEADER_SIZEOF + PCI_PM_SIZEOF + PCI_CAP_LIST_NEXT + 1]);
+    assert_int_equal(PCI_CAP_ID_VNDR,
+                     config_space.raw[PCI_STD_HEADER_SIZEOF + PCI_PM_SIZEOF + vsc[0]->vsc.size]);
+    assert_int_equal(0,
+                     config_space.raw[PCI_STD_HEADER_SIZEOF + PCI_PM_SIZEOF + vsc[0]->vsc.size + PCI_CAP_LIST_NEXT]);
+    assert_int_equal(vsc[1]->vsc.size,
+                     config_space.raw[PCI_STD_HEADER_SIZEOF + PCI_PM_SIZEOF + vsc[0]->vsc.size + PCI_CAP_LIST_NEXT + 1]);
+
+    /*
```

Comment:
I'm not following this FIXME, didn't we just do that?

Reply:
right